core/src/main/scala/com/eharmony/spotz/optimizer/grid/GridSearch.scala

```diff
@@ -0,0 +1,148 @@
+    val firstLoss = objective(firstPoint)
+    val currentTime = DateTime.now()
+
+    val gridSearchContext = GridSearchContext(
+      bestPointSoFar = firstPoint,
+      bestLossSoFar = firstLoss,
+      startTime = startTime,
+      currentTime = currentTime,
+      elapsedTime = new Duration(startTime, currentTime),
+      trialsSoFar = 1L,
+      optimizerFinished = 1L >= space.size)
+
+    // Last three arguments maintain the best point and loss and the trial count
+    gridSearch(objective, space, reducer, gridSearchContext)
+  }
+
+  @tailrec
+  private def gridSearch(objective: Objective[P, L],
+                         space: Grid[P],
+                         reducer: Reducer[(P, L)],
+                         gsc: GridSearchContext[P, L])
+                        (implicit c: ClassTag[P], p: ClassTag[L]): GridSearchResult[P, L] = {
+
+    info(s"Best point and loss after ${gsc.trialsSoFar} trials and ${DurationUtils.format(gsc.elapsedTime)} : ${gsc.bestPointSoFar} loss: ${gsc.bestLossSoFar}")
+
+    stopStrategy.shouldStop(gsc) match {
+
+      case true =>
+        // Base case, end recursion, return the result
+        GridSearchResult[P, L](
+          bestPoint = gsc.bestPointSoFar,
+          bestLoss = gsc.bestLossSoFar,
+          startTime = gsc.startTime,
+          endTime = gsc.currentTime,
+          elapsedTime = gsc.elapsedTime,
+          totalTrials = gsc.trialsSoFar)
+
+      case false =>
+        val currentTime = DateTime.now()
+        val elapsedTime = new Duration(gsc.startTime, currentTime)
+
+
+        val batchSize = scala.math.min(space.size - gsc.trialsSoFar, trialBatchSize)
+
+        val (bestPoint, bestLoss) = reducer((gsc.bestPointSoFar, gsc.bestLossSoFar), bestGridPointAndLoss(gsc.trialsSoFar, batchSize, objective, space, reducer))
+        val trialsSoFar = gsc.trialsSoFar + batchSize
+
+        val gridSearchContext = GridSearchContext(
+          bestPointSoFar = bestPoint,
+          bestLossSoFar = bestLoss,
+          startTime = gsc.startTime,
+          currentTime = currentTime,
+          elapsedTime = elapsedTime,
+          trialsSoFar = trialsSoFar,
+          optimizerFinished = trialsSoFar >= space.size)
+
+        gridSearch(objective, space, reducer, gridSearchContext)
+    }
+  }
+}
+
+case class GridSearchContext[P, L](
+    bestPointSoFar: P,
+    bestLossSoFar: L,
+    startTime: DateTime,
+    currentTime: DateTime,
+    elapsedTime: Duration,
+    trialsSoFar: Long,
+    optimizerFinished: Boolean) extends OptimizerState[P, L]
+
+/**
+  * Grid search backed by parallel collections.
+  *
+  * @param trialBatchSize default batch size specifying number of points to batch process
+  * @param ord Ordering for point type P
+  * @param factory a function that takes a Map of sampled hyper parameter values and instanties the point P
+  * @tparam P point type representation
+  * @tparam L loss type representation
+  */
+class ParGridSearch[P, L](
+    trialBatchSize: Int = 1000000,
```

Comment:
This controls the the number of trials that are processed per batch iteration.  At the time I set this value, the Branin function was being used as the objective.  Considering how quickly the objective function executes, this batch size can be tuned to be very high without any significant issue.  So it was arbitrarily chosen to be something high.

However, there are some impacts of setting this too low or too high.  Tuning the batch size parameter directly affects how long it takes a batch to complete.  The more trials to be passed to the objective function, the longer it takes a batch to complete, and the longer it takes to send an update back to the driver.  For objective functions that execute quickly, ie. microsecond duration, setting the batch trial size to something high isn't a huge deal.  For VW objective functions which take a few seconds to execute, this is many orders of magnitude higher and thusly much slower.  Setting the parameter too low in the case of the Branin function will actually incur more constant time overhead from Spark when returning the best point and loss back to the driver, ie. going back to the driver getting more points, running another batch, return best point and loss, rinse and repeat many many times.

That's the reason adaptive batch sizing is needed.  Without adaptive batch sizing in place, the takeaway is for fast objective functions, tune the batch trial size high.  For slower objective functions, tune the batch size lower.

Why not just parallelize all the points in one single batch and do one update to the driver?  Frequent intermediate updates to driver containing the best point and loss for that batch are good.  The best point and loss are logged so that user can see some progress.  Also, after giving control back to the driver allows it to update the state of the search, check stopping criteria, and also save out the search state in the case of a crash or intentional kill.